src/Aspire.Hosting/Dcp/DcpKubernetesClient.cs

```diff
@@ -0,0 +1,111 @@
+    /// <param name="version">The API version of the Kubernetes resource.</param>
+    /// <param name="plural">The plural name (API kind) of the Kubernetes resource, e.g. "executables".</param>
+    /// <param name="name">The name of the Kubernetes resource to use for sub-resource read operation.</param>
+    /// <param name="subResource">The sub-resource to read from the Kubernetes resource.</param>
+    /// <param name="namespaceParameter">The namespace of the Kubernetes resource.
+    /// If null or empty, the resource is assumed to be non-namespaced.</param>
+    /// <param name="queryParams">Optional query parameters to append to the request URL.</param>
+    /// <param name="cancellationToken">A token to monitor for cancellation requests.</param>
+    public async Task<HttpOperationResponse<Stream>> ReadSubResourceAsStreamAsync(
+        string group,
+        string version,
+        string plural,
+        string name,
+        string subResource,
+        string? namespaceParameter,
+        IReadOnlyCollection<(string name, string value)>? queryParams = null,
+        CancellationToken cancellationToken = default
+    )
+    {
+        ArgumentException.ThrowIfNullOrWhiteSpace(group, nameof(group));
+        ArgumentException.ThrowIfNullOrWhiteSpace(version, nameof(version));
+        ArgumentException.ThrowIfNullOrWhiteSpace(plural, nameof(plural));
+        ArgumentException.ThrowIfNullOrWhiteSpace(name, nameof(name));
+        ArgumentException.ThrowIfNullOrWhiteSpace(subResource, nameof(subResource));
+
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        cts.CancelAfter(HttpClientTimeout);
+        cancellationToken = cts.Token;
+
+        string url;
+        if (string.IsNullOrEmpty(namespaceParameter))
+        {
+            url = $"apis/{group}/{version}/{plural}/{name}/{subResource}";
+        }
+        else
+        {
+            url = $"apis/{group}/{version}/namespaces/{namespaceParameter}/{plural}/{name}/{subResource}";
+        }
+
+        var q = new QueryBuilder();
+        if (queryParams != null)
+        {
+            foreach (var (param, paramVal) in queryParams)
+            {
+                q.Append(param, paramVal);
+            }
+        }
+        url += q.ToString();
+
+        const IReadOnlyDictionary<string, IReadOnlyList<string>>? customHeaders = null;
+        var httpResponse = await SendRequest<object?>(url, HttpMethod.Get, customHeaders, body: null, cancellationToken).ConfigureAwait(false);
+        var httpRequest = httpResponse.RequestMessage;
+        var result = new HttpOperationResponse<Stream>()
+        {
+            Request = httpRequest,
+            Response = httpResponse,
+            Body = await httpResponse.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false)
+        };
+        return result;
+    }
+
+    private sealed class QueryBuilder
```

Comment:
Can you use [QueryHelpers](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.webutilities.queryhelpers?view=aspnetcore-8.0) instead of a handmade class for building the query string?

Reply:
The QueryHelpers only works with string values. I wanted to encapsulate the logic for other types of parameters.